### PR TITLE
Particle Filter Implementation

### PR DESCRIPTION
**Pull request by Marwan99**

### CODE REVIEW

File: nav2_localization/src/plugins/solvers/particle_filter_base.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    quat.setRPY(0.0, 0.0, yaw);
+    temp_pose.orientation = tf2::toMsg(quat);
+
+    particles_.push_back(Particle(temp_pose, 1));
+  }
+
+  visualize_particles();
+}
+
+void ParticleFilter::resample()
```
Comment (khaled-98) [inline nav2_localization/src/plugins/solvers/particle_filter_base.cpp:78]: Add a comment to discuss if this the best place for this to be implemented

File: nav2_localization/src/plugins/solvers/particle_filter_base.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+{
+  std::vector<Particle> sampled_particles;
+
+  double particle_count_dtype = static_cast<double>(particles_count_);
+
+  std::random_device rand_device;
+  std::mt19937 gen(rand_device());
+  std::uniform_real_distribution<double> dist(0, 1 / particle_count_dtype);
+
+  double c = particles_[0].weight_;
```
Comment (khaled-98) [inline nav2_localization/src/plugins/solvers/particle_filter_base.cpp:88]: Please update c, u, r and m to clarify what they do

File: nav2_localization/src/plugins/solvers/particle_filter_base.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+{
+  double mean_x = 0, mean_y = 0, mean_yaw = 0;
+  double weights_sum = 0;
+
+  for (auto particle : particles_) {
+    mean_x += particle.weight_ * particle.pose_.position.x;
+    mean_y += particle.weight_ * particle.pose_.position.y;
+
+    double yaw = tf2::getYaw(particle.pose_.orientation);
+    mean_yaw += particle.weight_ * yaw;
```
Comment (khaled-98) [inline nav2_localization/src/plugins/solvers/particle_filter_base.cpp:138]: Please fix the mean yaw

**Review by khaled-98 — CHANGES_REQUESTED** [review]

File: nav2_localization/src/plugins/solvers/mcl_solver.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  // TODO(marwan99): Add more resampling throttling methods e.g. frequency.
+  // TODO(marwan99): Make dist threshold a parameter
+  bool do_resmaple = true;  // Set to false to enable throtelling 
+  double delta_x = prev_odom_.pose.pose.position.x - curr_odom.pose.pose.position.x;
+  double delta_y = prev_odom_.pose.pose.position.y - curr_odom.pose.pose.position.y;
+  double delta_yaw = angles::normalize_angle(
+    tf2::getYaw(prev_odom_.pose.pose.orientation) - 
+    tf2::getYaw(curr_odom.pose.pose.orientation));
+  if (delta_x * delta_x + delta_y * delta_y > 0.01 || delta_yaw > 0.25) {
```
Comment (khaled-98) [inline nav2_localization/src/plugins/solvers/mcl_solver.cpp:26]: Take abs value of delta_yaw

File: nav2_localization/include/nav2_localization/nav2_localization.hpp [hunk 5]
```diff
@@ -127,26 +129,37 @@ class LocalizationServer : public nav2_util::LifecycleNode
...
+    sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan,
+    nav_msgs::msg::Odometry::ConstSharedPtr odom);
 
   /**
    * @brief Callback when the scan is received
    * @param scan pointer to the received PointCloud2 message
+   * @param odom pointer to the received odometry message
    */
-  void scanReceived(sensor_msgs::msg::PointCloud2::ConstSharedPtr scan);
+  void sensorsReceived(
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply nav2_localization/include/nav2_localization/nav2_localization.hpp:141]: How about scanWithOdomRecieved or similar? And do something similar with  the laser callback. Just to accurately reflect whats happening.

File: nav2_localization/src/plugins/solvers/particle_filter_base.cpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & node,
+  SampleMotionModel::Ptr & motionSampler,
+  Matcher2d::Ptr & matcher)
+{
+  Solver::configure(node, motionSampler, matcher);
+
+  particle_cloud_pub_ = node_->create_publisher<nav2_msgs::msg::ParticleCloud>(
+    "particle_cloud", rclcpp::SensorDataQoS());
+
+  particles_poses_pub_ = node_->create_publisher<geometry_msgs::msg::PoseArray>(
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply nav2_localization/src/plugins/solvers/particle_filter_base.cpp:31]: Added comment in a different place.

**Review by khaled-98 — APPROVED** [review]

**Review by khaled-98 — APPROVED** [review]

**Review by khaled-98 — APPROVED** [review]

**Pull request merged**